### PR TITLE
The View Audit

### PR DESCRIPTION
**Pull request by park66665**
## About The Pull Request

- [x] First pass of `range` review (>200 usages)
- [x] Second pass of `range` review
- [x] Third pass of `range` review
- [x] Further review of `range` changes
- [ ] Thoroughly benchmark to see if `RANGE_TURFS` is strictly a better choice
- [x] First pass of `view` review (>150 usages)
- [x] Second pass of `view` review
- [x] Third pass of `view` review
- [x] Further review of `view` changes
- [x] `orange` review (c. 80 usages)
- [x] `oview` review (c. 70 usages)
- [x] `hearers` and `ohearers` review (20~30 usages)
- [x] `viewers` and `oviewers` review (c. 60 usages)
- [x] `/mob/living/simple_animal/hostile/proc/ListTargets()` (c. 50% reduction; needs more benchmarks, though.)
- [x] `get_hearers_in_view` (c. 30% reduction; the best one was chosen among 11 candidates)
- [x] `/mob/living/simple_animal/bot/proc/scan` (needs a *lot* of love)
- [ ] Optimize other hot procs

This reviews `view` and its friends. This drive kinda sorta started when maintainers were reviewing the performance and hot codes. This PR segues to `Life()` review a little bit, but, surely, that goal isn't getting attention.

### A really short introduction of `view`-esque procs.

BYOND has eight procs that can poll nearby things. `range`, `orange`, `view`, `oview`, `hearers`, `ohearers`, `viewers`, and `oviewers`. Read DM reference for their oddities and properties. The first four of them can return a *huge* list of atoms. `range` returns the most, while `oviewers` will return the least. The hierarchy of the number of returned atoms is roughly:

> range > orange ⪆ view > oview > (The uncrossable 4-dimensional wall) > hearers ≥ ohearers ≥ viewers ≥ oviewers

The cost hierarchy is roughly:

> range ≈ orange ≥ view ≈ oview ≫ hearers ≈ ohearers ≈ viewers ≈ oviewers

So, in a nutshell, `range` and `view` usage should be curtailed. Not every one of them can be substituted, of course. I know BYOND does a lot of magic to bring associated costs down, but when it comes to `for(var/something/something in range(range, src))` stuff, the inefficiency can go off the charts.

Here is a rough list of things that should be noted when changing the procs:
1. `range` sometimes can't be substituted. It currently is *the* way to indiscriminately poll every nearby atom.
2. `view` from most mobs and turfs also return barely usable atoms, like lighting objects or other invisible objects
3. `view`, `oview`, `viewers`, and `oviewers` account for luminosity and opacity. `hearers` and `ohearers` account for opacity. (think `dview`) `range` and `orange` account only for distance.
4. visibility is noncommutative.
5. mechas aren't mobs.
6. cameras are invisible mobs.

Another thing to note is helper procs; quite a lot of them exists, but only two of them are in active use: `get_hearers_in_view` (See #3338 and #3340) and `get_hear`. Other helper procs aren't very important, at least for me.

## Why It's Good For The Game

If this audit goes well, things will be a bit faster, and I hope that this audit will be an example for future coders. Don't abuse `range`.

## Changelog
:cl:
code: range, view and their related high-cost procs' usages have been audited
fix: above audit should improve performance, but there might be bugs. Please report bugs to GitHub.
/:cl:

### CODE REVIEW

File: code/game/objects/effects/anomalies.dm [hunk 1]
```diff
@@ -92,14 +92,14 @@
...
 		if(!O.anchored)
 			step_towards(O,src)
-	for(var/mob/living/M in range(0, src))
+	for(var/mob/living/M in ohearers(0, src))
 		gravShock(M)
-	for(var/mob/living/M in orange(4, src))
+	for(var/mob/living/M in orange(4, get_turf(src)))
 		if(!M.mob_negates_gravity())
 			step_towards(M,src)
 	for(var/obj/O in range(0,src))
```
Comment (ike709) [inline code/game/objects/effects/anomalies.dm]:
```suggestion
	for(var/obj/O in get_turf(src))
```

File: code/game/objects/effects/effect_system/effects_foam.dm [hunk 2]
```diff
@@ -146,7 +146,7 @@
 		if(lifetime % reagent_divisor)
 			reagents.reaction(O, VAPOR, fraction)
 	var/hit = 0
-	for(var/mob/living/L in range(0,src))
+	for(var/mob/living/L in hearers(0,src))
```
Comment (ike709) [inline code/game/objects/effects/effect_system/effects_foam.dm]:
```suggestion
	for(var/mob/living/L in get_turf(src))
```

**Comment by ike709** [comment]
Reminder that `Adjacent()` can likely be used in place of any checks for something being within 1 turf's distance.

**Comment by PowerfulBacon** [comment]
good shit

File: code/datums/brain_damage/severe.dm [hunk 3]
```diff
@@ -167,9 +167,7 @@
 /datum/brain_trauma/severe/monophobia/proc/check_alone()
 	if(HAS_TRAIT(owner, TRAIT_BLIND))
 		return TRUE
-	for(var/mob/M in oview(owner, 7))
-		if(!isliving(M)) //ghosts ain't people
-			continue
+	for(var/mob/living/M in oview(7, owner))
```
Comment (ike709) [inline code/datums/brain_damage/severe.dm:170]: Hnnnnnnnng this proc is called in `on_life()`

:weary:

**Comment by park66665** [comment]
This PR still needs another pass of my own review before un-drafting. There shouldn't be another part of View Audit.

File: code/modules/mob/living/simple_animal/hostile/retaliate/retaliate.dm [hunk 4]
```diff
@@ -21,9 +21,7 @@
 	return see
 
 /mob/living/simple_animal/hostile/retaliate/proc/Retaliate()
-	var/list/around = oview(vision_range, src)
-
-	for(var/atom/movable/A as obj|mob in around)
+	for(var/atom/movable/A as obj|mob in oview(vision_range, src))
```
Comment (ike709) [inline code/modules/mob/living/simple_animal/hostile/retaliate/retaliate.dm:24]: hmmmm

**Comment by park66665** [comment]
This PR seems subjectively acceptably polished after three reviews. I'll un-draft this PR.

File: code/modules/mob/living/simple_animal/hostile/retaliate/retaliate.dm (hunk 4 above)
Comment (park66665) [reply code/modules/mob/living/simple_animal/hostile/retaliate/retaliate.dm:24]: blame mechas for forcing multiple loops to be `[o]view` that could've been `[o]viewers` or `[o]hearers`

**Review by ike709 — DISMISSED** [review]
_Seems_ fine?

**Review by ike709 — DISMISSED** [review]

**Comment by ike709** [comment]
Since nobody else seems willing to review this, just fix the conflicts one last time and I'll yolomerge.

**Comment by ike709** [comment]
YOLO

**Pull request merged**